trace-normalization/src/normalize_utils.rs

```diff
@@ -1,332 +1,367 @@
+    let mut write_cursor = 0;
+    let mut last_written_char = 0;
+    loop {
+        if read_cursor >= bytes.len() {
+            break;
         }
-    };
-
-    while i < name.len() {
-        if is_alpha_num(char_vec[i]) {
-            result.push(char_vec[i]);
-            last_char = char_vec[i];
-        } else if char_vec[i] == '.' {
-            // we skipped all non-alpha chars up front so we have seen at least one
-            if last_char == '_' {
-                // overwrite underscores that happen before periods
-                result.replace_range((result.len() - 1)..(result.len()), ".");
-                last_char = '.'
-            } else {
-                result.push('.');
-                last_char = '.';
+        match (bytes[read_cursor], last_written_char) {
+            (b @ (b'a'..=b'z' | b'A'..=b'Z' | b'0'..=b'9'), _) => {
+                bytes[write_cursor] = b;
+                last_written_char = b;
+            }
+            // If we've written a _ last, replace it with a .
+            (b'.', b'_') => {
+                // This safe because the first character is alpha so
+                // we don't go back to the beginning
+                write_cursor -= 1;
+                bytes[write_cursor] = b'.';
+                last_written_char = b'.'
+            }
+            // If we've written a _ or a . last, do nothing
+            (_, b'_' | b'.') => {}
+            (b @ (b'_' | b'.'), _) => {
+                bytes[write_cursor] = b;
+                last_written_char = b;
             }
-        } else {
-            // we skipped all non-alpha chars up front so we have seen at least one
-            if last_char != '.' && last_char != '_' {
-                result.push('_');
-                last_char = '_';
+            // Otherwise write _ instead of any non conforming char
+            (_, _) => {
+                bytes[write_cursor] = b'_';
+                last_written_char = b'_';
             }
         }
-        i += 1;
+        write_cursor += 1;
+        read_cursor += 1;
     }
-
-    if last_char == '_' {
-        result.remove(result.len() - 1);
+    if last_written_char == b'_' {
+        write_cursor -= 1;
     }
-    Ok(result)
-}
+    bytes.truncate(write_cursor);
 
-pub(crate) fn is_alpha(c: char) -> bool {
-    c.is_ascii_lowercase() || c.is_ascii_uppercase()
+    // We only wrote ascii chars, so bytes is guaranteed to be valid utf8
+    *name = String::from_utf8(bytes).unwrap();
 }
 
-pub(crate) fn is_alpha_num(c: char) -> bool {
-    is_alpha(c) || c.is_ascii_digit()
+// truncate_utf8 truncates the given string to make sure it uses less than limit bytes.
+// If the last character is a utf8 character that would be split, it removes it
+// entirely to make sure the resulting string is not broken.
+pub(crate) fn truncate_utf8(s: &mut String, limit: usize) {
+    let boundary = crate::utf8_helpers::floor_char_boundary(s, limit);
+    s.truncate(boundary);
```

Comment:
Is `s.truncate(...)` a no-op if the boundary is the length of `s`?

Reply:
[Almost](https://doc.rust-lang.org/src/alloc/string.rs.html#1406) it is checking if the boundary is higher or equal than the length. And since all the functions used are inline, the compiler might be reusing the call to `len` from `floor_char_boundary`.